mlflow/langchain/databricks_dependencies.py

```diff
@@ -0,0 +1,185 @@
+
+
+def _assign_value_or_append_to_list(d, key, value):
+    if key in d:
+        if isinstance(d[key], list):
+            d[key].append(value)
+        else:
+            d[key] = [d[key], value]
+    else:
+        d[key] = value
+    return d
+
+
+def _extract_databricks_dependencies_from_retriever(retriever, dependency_dict) -> Dict[str, str]:
+    import langchain
+
+    if hasattr(retriever, "vectorstore") and hasattr(retriever.vectorstore, "embeddings"):
+        vectorstore = retriever.vectorstore
+        embeddings = vectorstore.embeddings
+
+        if isinstance(vectorstore, langchain.vectorstores.DatabricksVectorSearch):
+            index = vectorstore.index
+            _assign_value_or_append_to_list(
+                dependency_dict, _DATABRICKS_VECTOR_SEARCH_INDEX_NAME_KEY, index.name
+            )
+            _assign_value_or_append_to_list(
+                dependency_dict, _DATABRICKS_VECTOR_SEARCH_ENDPOINT_NAME_KEY, index.endpoint_name
+            )
+
+        if isinstance(embeddings, langchain.embeddings.DatabricksEmbeddings):
+            _assign_value_or_append_to_list(
+                dependency_dict, _DATABRICKS_EMBEDDINGS_ENDPOINT_NAME_KEY, embeddings.endpoint
+            )
+        elif (
+            hasattr(vectorstore, "_is_databricks_managed_embeddings")
+            and callable(getattr(vectorstore, "_is_databricks_managed_embeddings"))
+            and vectorstore._is_databricks_managed_embeddings()
+        ):
+            _assign_value_or_append_to_list(
+                dependency_dict,
+                _DATABRICKS_EMBEDDINGS_ENDPOINT_NAME_KEY,
+                "_is_databricks_managed_embeddings",
+            )
+
+    return dependency_dict
+
+
+def _extract_databricks_dependencies_from_llm(llm, dependency_dict) -> Dict[str, str]:
+    import langchain
+
+    if isinstance(llm, langchain.llms.Databricks):
+        _assign_value_or_append_to_list(
+            dependency_dict, _DATABRICKS_LLM_ENDPOINT_NAME_KEY, llm.endpoint_name
+        )
+    return dependency_dict
+
+
+def _extract_databricks_dependencies_from_chat_model(chat_model, dependency_dict) -> Dict[str, str]:
+    import langchain
+
+    if isinstance(chat_model, langchain.chat_models.ChatDatabricks):
+        _assign_value_or_append_to_list(
+            dependency_dict, _DATABRICKS_CHAT_ENDPOINT_NAME_KEY, chat_model.endpoint
+        )
+    return dependency_dict
+
+
+def _extract_dependency_dict_from_lc_model(lc_model, dependency_dict) -> Dict[str, str]:
+    """
+    This function contains the logic to examine a non-Runnable component of a langchain model.
+    The logic here does not cover all legacy chains. If you need to support a custom chain,
+    you need to monkey patch this function.
+    """
+    if lc_model is None:
+        return dependency_dict
+
+    # leaf node
+    dependency_dict = _extract_databricks_dependencies_from_chat_model(lc_model, dependency_dict)
```

Comment:
I would prefer a type check of lc_model here, and inside each function I think we should raise Exception if the type mismatches

Reply:
Hey, just curious - do you think the type checks on lines 29, 38, 59, and 69 cover everything we need? Is there anything else you'd suggest adding in for a bit more security?

Reply:
Is there any concerns? They look reasonable and cover what we want to log 😆

Reply:
> I would prefer a type check of lc_model here, and inside each function I think we should raise Exception if the type mismatches

So the suggested "type check here and Exception" won't be needed, right?

Reply:
Oh yes, it's not necessary, it just looks more reasonable that we only extract the dependencies for corresponding model types :) But those functions are internal anyways, so I'm fine with what it is now as well